### PR TITLE
Better handling existing secret

### LINKED ISSUE #48: Better handling for existing secrets
Currently we use `Update` instead of `Create` to avoid dealing with existing secrets. However we need to handle this more properly.

Relevant code is here: https://github.com/istio/auth/blob/master/controller/secret.go#L121

### PR DESCRIPTION
**Pull request by lookuptable**
Fix #48

### CODE REVIEW

File: controller/secret.go [hunk 1]
```diff
@@ -36,10 +36,12 @@ const secretNamePrefix = "istio."
...
 type SecretController struct {
-	ca           certmanager.CertificateAuthority
-	core         corev1.CoreV1Interface
-	saController cache.Controller
-	saStore      cache.Store
+	ca             certmanager.CertificateAuthority
+	core           corev1.CoreV1Interface
+	saController   cache.Controller
+	saStore        cache.Store
+	scrtController cache.Controller
```
Comment (wattli) [inline controller/secret.go]: Add some comments about this variable and the one below?

File: controller/secret.go [hunk 2]
```diff
@@ -62,20 +64,31 @@ func NewSecretController(ca certmanager.CertificateAuthority, core corev1.CoreV1
...
 func (sc *SecretController) Run(stopCh chan struct{}) {
+	go sc.scrtController.Run(stopCh)
 	go sc.saController.Run(stopCh)
 	<-stopCh
 }
 
 func (sc *SecretController) addFunc(obj interface{}) {
 	acct := obj.(*v1.ServiceAccount)
-	sc.createSecret(acct.GetName(), acct.GetNamespace())
+	sc.createOrUpdateSecret(acct.GetName(), acct.GetNamespace())
```
Comment (wattli) [inline controller/secret.go]: Nit: maybe call it upsertSecret?

File: controller/secret.go (hunk 1 above)
Comment (wattli) [inline controller/secret.go]: Or just call it 'secretController' to be consistent with secretlw below.

File: controller/secret.go [hunk 3]
```diff
@@ -49,7 +51,7 @@ func NewSecretController(ca certmanager.CertificateAuthority, core corev1.CoreV1
 		core: core,
 	}
 
-	lw := &cache.ListWatch{
+	salw := &cache.ListWatch{
```
Comment (wattli) [inline controller/secret.go]: saLw? To make it more readable. Same for secretLw.

File: controller/secret.go [hunk 4]
```diff
@@ -62,20 +64,31 @@ func NewSecretController(ca certmanager.CertificateAuthority, core corev1.CoreV1
...
+
+	secretlw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return core.Secrets(metav1.NamespaceAll).List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return core.Secrets(metav1.NamespaceAll).Watch(options)
+		},
+	}
+	c.scrtStore, c.scrtController = cache.NewInformer(secretlw, &v1.Secret{}, time.Minute, cache.ResourceEventHandlerFuncs{})
```
Comment (wattli) [inline controller/secret.go]: So you will have a followup PR for this handlerFuncs? If so, please leave a TODO here.

File: controller/secret.go (hunk 1 above)
Comment (lookuptable) [inline controller/secret.go]: Done

File: controller/secret.go (hunk 3 above)
Comment (lookuptable) [inline controller/secret.go]: Done

File: controller/secret.go (hunk 2 above)
Comment (lookuptable) [inline controller/secret.go]: Done

**Pull request merged**